src/dbos-executor.ts

```diff
@@ -907,34 +927,22 @@ export class DBOSExecutor implements DBOSExecutorContext {
           wCtxt.span.setAttribute('cached', true);
           wCtxt.span.setStatus({ code: SpanStatusCode.OK });
         } else if (err instanceof DBOSWorkflowCancelledError) {
+          wCtxt.span.setStatus({ code: SpanStatusCode.ERROR, message: err.message });
           internalStatus.error = err.message;
-          internalStatus.status = StatusString.CANCELLED;
-
-          if (!this.isDebugging) {
-            await this.systemDatabase.cancelWorkflow(workflowID);
+          if (err.workflowID === workflowID) {
+            internalStatus.status = StatusString.CANCELLED;
+            if (!this.isDebugging) {
+              await this.systemDatabase.cancelWorkflow(workflowID);
+            }
+            this.logger.info(`Cancelled workflow ${workflowID}`);
+            throw err;
+          } else {
+            const e = new DBOSAwaitedWorkflowCancelledError(workflowID);
+            await handleWorkflowError(e as Error);
```

Comment:
Why is this needed?

Reply:
We need to record the error on this path too -- otherwise the wf status will be `CANCELLED`. This is related to the comment you astutely made here https://github.com/dbos-inc/dbos-transact-ts/pull/903#discussion_r2089897918